File: flutter_app/lib/presentation/chat/chat_view_model.dart
import 'package:flutter/material.dart';
import 'package:firebase_auth/firebase_auth.dart' as auth;
import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:uuid/uuid.dart';

import '../../domain/model/data_models.dart';

class ChatViewModel extends ChangeNotifier {
  final auth.FirebaseAuth _auth = auth.FirebaseAuth.instance;
  final FirebaseFirestore _firestore = FirebaseFirestore.instance;

  List<ChatMessage> _messages = [];
  List<ChatMessage> get messages => _messages;

  String get currentUserId => _auth.currentUser?.uid ?? "";
  String _currentUserName = "User";
  String get currentUserName => _currentUserName;

  ChatViewModel() {
    _loadUserName();
  }

  Future<void> _loadUserName() async {
    if (currentUserId.isNotEmpty) {
      try {
        final doc = await _firestore.collection("users").doc(currentUserId).get();
        if (doc.exists) {
          _currentUserName = doc.data()?['name'] ?? "User";
          notifyListeners();
        }
      } catch (e) {
        debugPrint(e.toString());
      }
    }
  }

  void loadMessages(String bookingId) {
    _firestore.collection("chats")
        .where("bookingId", isEqualTo: bookingId)
        .snapshots()
        .listen((snapshot) {
      final list = snapshot.docs.map((doc) => ChatMessage.fromMap(doc.data())).toList();
      list.sort((a, b) => b.timestamp.compareTo(a.timestamp)); // Most recent first for ListView.builder reversed
      _messages = list;
      notifyListeners();
    });
  }

  Future<void> sendMessage(String bookingId, String text, {String? imageUrl}) async {
    if (text.trim().isEmpty && (imageUrl == null || imageUrl.isEmpty)) return;
    
    final msgId = const Uuid().v4();
    final chatMsg = ChatMessage(
      id: msgId,
      bookingId: bookingId,
      senderId: currentUserId,
      senderName: _currentUserName,
      text: text.trim(),
      imageUrl: imageUrl ?? "",
      timestamp: DateTime.now().millisecondsSinceEpoch,
    );

    try {
      await _firestore.collection("chats").doc(msgId).set(chatMsg.toMap());
    } catch (e) {
      debugPrint(e.toString());
    }
  }
}

File: app/src/main/java/com/kina/care/presentation/chat/ChatViewModel.kt
package com.kina.care.presentation.chat

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.firestore.FirebaseFirestore
import com.kina.care.domain.model.ChatMessage
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import kotlinx.coroutines.tasks.await
import java.util.UUID

class ChatViewModel : ViewModel() {
    private val firestore = FirebaseFirestore.getInstance()
    private val auth = FirebaseAuth.getInstance()

    private val _messages = MutableStateFlow<List<ChatMessage>>(emptyList())
    val messages: StateFlow<List<ChatMessage>> = _messages.asStateFlow()

    val currentUserId = auth.currentUser?.uid ?: ""
    var currentUserName = "User"

    init {
        // ලොග් වී ඇති කෙනාගේ නම ලබාගැනීම
        viewModelScope.launch {
            if (currentUserId.isNotEmpty()) {
                try {
                    val doc = firestore.collection("users").document(currentUserId).get().await()
                    currentUserName = doc.getString("name") ?: "User"
                } catch(e: Exception){}
            }
        }
    }

    fun loadMessages(bookingId: String) {
        firestore.collection("chats")
            .whereEqualTo("bookingId", bookingId)
            .addSnapshotListener { snapshot, error ->
                if (error != null) return@addSnapshotListener
                if (snapshot != null) {
                    val list = snapshot.documents.mapNotNull { it.toObject(ChatMessage::class.java) }
                        .sortedBy { it.timestamp } // පරණ මැසේජ් උඩින්, අලුත් ඒවා යටින්
                    _messages.value = list
                }
            }
    }

    fun sendMessage(bookingId: String, text: String, imageUrl: String? = null) {
        if (text.isBlank() && imageUrl.isNullOrBlank()) return
        viewModelScope.launch {
            val msgId = UUID.randomUUID().toString()
            val chatMsg = ChatMessage(
                id = msgId,
                bookingId = bookingId,
                senderId = currentUserId,
                senderName = currentUserName,
                text = text.trim(),
                imageUrl = imageUrl,
                timestamp = System.currentTimeMillis()
            )
            try {
                firestore.collection("chats").document(msgId).set(chatMsg).await()
            } catch(e: Exception) { e.printStackTrace() }
        }
    }
}